doc/designs/quic-design/quic-fault-injector.md

```diff
@@ -0,0 +1,566 @@
+                                                   void *cbarg);
+
+int ossl_quic_fault_set_packet_cipher_listener(OSSL_QUIC_FAULT *fault,
+                                    ossl_quic_fault_on_packet_cipher_cb pciphercb,
+                                    void *picphercbarg);
+
+/*
+ * Enable tests to listen for datagrams being sent
+ */
+typedef int (*ossl_quic_fault_on_datagram_cb)(OSSL_QUIC_FAULT *fault,
+                                              BIO_MSG *m,
+                                              size_t stride,
+                                              void *cbarg);
+
+int ossl_quic_fault_set_datagram_listener(OSSL_QUIC_FAULT *fault,
+                                          ossl_quic_fault_on_datagram_cb datagramcb,
+                                          void *datagramcbarg);
+
+/*
+ * To be called from a datagram_listener callback. The datagram buffer is over
+ * allocated, so this just changes the logical size and never changes the actual
+ * address of the buffer. This will fail if a large resize is attempted that
+ * exceeds the over allocation.
+ */
+int ossl_quic_fault_resize_datagram(OSSL_QUIC_FAULT *fault, size_t newlen);
+
+````
+
+Example Tests
+-------------
+
+This section provides some example tests to illustrate how the Fault Injector
+might be used to create tests.
+
+### Unknown Frame Test
+
+An example test showing a server sending a frame of an unknown type to the
+client:
+
+```` C
+/*
+ * Test that adding an unknown frame type is handled correctly
+ */
+static int add_unknown_frame_cb(OSSL_QUIC_FAULT *fault, QUIC_PKT_HDR *hdr,
+                                unsigned char *buf, size_t len, void *cbarg)
+{
+    size_t done = 0;
+
+    /*
+     * There are no "reserved" frame types which are definitately safe for us
+     * to use for testing purposes - but we just use the highest possible
+     * value (8 byte length integer) and with no payload bytes
+     */
+    unsigned char unknown_frame[] = {
+        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff
+    };
+
+    /* We only ever add the unknown frame to one packet */
+    if (done)
+        return 1;
+    done++;
+
+    /* Extend the size of the packet by the size of the new frame */
+    if (!TEST_true(ossl_quic_fault_resize_plain_packet(fault,
+                                                       len + sizeof(unknown_frame))))
+        return 0;
+
+    /*
+     * We prepend the new frame to the start of the packet. We add it to the
+     * start rather than the end because stream frames that are already in the
+     * packet may not have an explicit length, and instead may just extend to
+     * the end of the packet. We could fix-up such frames to have an explicit
+     * length and add our new frame after it. But it is probably simpler just to
+     * add it to the beginning of the packet. This means moving the existing
+     * packet data.
+     */
+    memmove(buf + sizeof(unknown_frame), buf, len);
```

Comment:
This should be in a helper function.  Insert frame at start, insert frame before, insert frame after, ...

Reply:
I made it into a helper function as suggested.